app/saml.go

```diff
@@ -172,3 +175,91 @@ func (a *App) GetSamlCertificateStatus() *model.SamlCertificateStatus {
 
 	return status
 }
+
+func (a *App) GetSamlMetadataFromIdp(idpMetadataUrl string) (*model.SamlMetadataResponse, *model.AppError) {
+	if a.Saml == nil {
+		err := model.NewAppError("GetSamlMetadataFromIdp", "api.admin.saml.not_available.app_error", nil, "", http.StatusNotImplemented)
+		return nil, err
+	}
+
+	idpMetadata, err := a.FetchSamlMetadataFromIdp(idpMetadataUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	data := &model.SamlMetadataResponse{}
+	data.IdpDescriptorUrl = idpMetadata.EntityID
+
+	if idpMetadata.IDPSSODescriptors == nil || len(idpMetadata.IDPSSODescriptors) == 0 {
+		err := model.NewAppError("GetSamlMetadataFromIdp", "api.admin.saml.invalid_xml_missing_idpssodescriptors.app_error", nil, "", http.StatusInternalServerError)
+		return nil, err
+	}
+
+	idpSSODescriptor := idpMetadata.IDPSSODescriptors[0]
+	if idpSSODescriptor.SingleSignOnServices == nil || len(idpSSODescriptor.SingleSignOnServices) == 0 {
+		err := model.NewAppError("GetSamlMetadataFromIdp", "api.admin.saml.invalid_xml_missing_ssoservices.app_error", nil, "", http.StatusInternalServerError)
+		return nil, err
+	}
+
+	data.IdpUrl = idpSSODescriptor.SingleSignOnServices[0].Location
+	if idpSSODescriptor.SSODescriptor.RoleDescriptor.KeyDescriptors == nil || len(idpSSODescriptor.SSODescriptor.RoleDescriptor.KeyDescriptors) == 0 {
+		err := model.NewAppError("GetSamlMetadataFromIdp", "api.admin.saml.invalid_xml_missing_keydescriptor.app_error", nil, "", http.StatusInternalServerError)
+		return nil, err
+	}
+	keyDescriptor := idpSSODescriptor.SSODescriptor.RoleDescriptor.KeyDescriptors[0]
+	data.IdpPublicCertificate = keyDescriptor.KeyInfo.X509Data.X509Certificate.Cert
+
+	return data, nil
+}
+
+func (a *App) FetchSamlMetadataFromIdp(url string) (*model.EntityDescriptor, *model.AppError) {
+	resp, err := http.Get(url)
+	if err != nil || resp.StatusCode != http.StatusOK {
+		return nil, model.NewAppError("FetchSamlMetadataFromIdp", "app.admin.saml.invalid_response_from_idp.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+	defer resp.Body.Close()
+
+	bodyXML, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, model.NewAppError("FetchSamlMetadataFromIdp", "app.admin.saml.failure_read_response_body_from_idp.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	entityDescriptor := model.EntityDescriptor{}
+	err = xml.Unmarshal(bodyXML, &entityDescriptor)
+	if err != nil {
+		return nil, model.NewAppError("FetchSamlMetadataFromIdp", "app.admin.saml.failure_decode_metadata_xml_from_idp.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	return &entityDescriptor, nil
```

Comment:
I wonder if it would make sense to split `FetchSamlMetadataFromIdp` and `GetSamlMetadataFromIdp` into one method that simply returns response the payload from a URL, and another method that takes the response string and unmarshals it into a `model.SamlMetadataResponse`? Right now they both have knowledge of the URL and unmarshaling process.

Reply:
Makes sense, the two steps should more cleanly separated now.